modules/distribution-zones/src/test/java/org/apache/ignite/internal/distributionzones/DistributionZoneManagerAlterFilterTest.java

```diff
@@ -0,0 +1,340 @@
+        assertDataNodesFromManager(
+                distributionZoneManager,
+                DEFAULT_ZONE_ID,
+                Set.of(C, D).stream().map(ClusterNode::name).collect(Collectors.toSet()),
+                10_000
+        );
+    }
+
+    /**
+     * Tests that altering of a filter affects only scale up timers and only added nodes.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    void testAlterFilterDoNotAffectScaleDown() throws Exception {
+        preparePrerequisites(IMMEDIATE_TIMER_VALUE, 10000);
+
+        topology.putNode(D);
+
+        assertNull(distributionZoneManager.zonesState().get(ZONE_ID).scaleDownTask());
+
+        topology.removeNodes(Set.of(C));
+
+        // Check that scale down task was scheduled.
+        assertTrue(waitForCondition(() -> distributionZoneManager.zonesState().get(ZONE_ID).scaleDownTask() != null, 10_000));
+
+        String newFilter = "$[?(@.region == 'CN')]";
+
+        distributionZoneManager.alterZone(ZONE_NAME,
+                new DistributionZoneConfigurationParameters.Builder(ZONE_NAME)
+                        .filter(newFilter)
+                        .build()
+        ).get(10_000, TimeUnit.MILLISECONDS);
+
+        // TODO: https://issues.apache.org/jira/browse/IGNITE-19506 change this to the causality versioned call to dataNodes.
+        // Node C is still in data nodes because altering a filter triggers only immediate scale up.
+        assertDataNodesFromManager(
+                distributionZoneManager,
+                ZONE_ID,
+                Set.of(C, D).stream().map(ClusterNode::name).collect(Collectors.toSet()),
+                10_000
+        );
+
+        // Check that scale down task is still scheduled.
+        assertNotNull(distributionZoneManager.zonesState().get(ZONE_ID).scaleUpTask());
+
+        // Alter zone so we could check that node C is removed from data nodes.
+        distributionZoneManager.alterZone(
+                ZONE_NAME,
+                new DistributionZoneConfigurationParameters.Builder(ZONE_NAME)
+                        .dataNodesAutoAdjustScaleUp(INFINITE_TIMER_VALUE)
+                        .dataNodesAutoAdjustScaleDown(IMMEDIATE_TIMER_VALUE)
+                        .filter(newFilter)
+                        .build()
+        ).get(10_000, TimeUnit.MILLISECONDS);
+
+        assertDataNodesFromManager(
+                distributionZoneManager,
+                ZONE_ID,
+                Set.of(D).stream().map(ClusterNode::name).collect(Collectors.toSet()),
+                10_000
+        );
+    }
+
+    /**
+     * Tests that node, that was added after altering of a filter (meaning, that immediate scale up was triggered), is added to data nodes
+     * only after altering and corresponding scale up.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    void testNodeAddedWhileAlteringFilter() throws Exception {
+        preparePrerequisites(10000, INFINITE_TIMER_VALUE);
```

Comment:
Here and there, 10000 -> 10_000 -> constant.

Reply:
done